source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,762 @@
+    // Case 2: first operand is a constant created with OpConstantNull.
+    return BuildInstructionAndAddToModule(CreateConst(GetType(inst), {}), pos);
+  } else {
+    // No other valid cases.
+    return nullptr;
+  }
+}
+
+ir::Instruction* FoldSpecConstantOpAndCompositePass::DoVectorShuffle(
+    ir::Module::inst_iterator* pos) {
+  ir::Instruction* inst = &**pos;
+  analysis::Vector* result_vec_type = GetType(inst)->AsVector();
+  assert(inst->NumInOperands() - 1 > 2 &&
+         "OpSpecConstantOp DoVectorShuffle instruction requires more than 2 "
+         "operands (2 vector ids and at least one literal operand");
+  assert(result_vec_type &&
+         "The result of VectorShuffle must be of type vector");
+
+  // A temporary buffer of new null constants to be build in case any of the
+  // vector operands are declared with OpConstantNull.
+  std::vector<std::unique_ptr<analysis::Constant> > null_constants_to_build;
+
+  // Get a concatenated vector of scalar constants. The vector should be built
+  // with the components from the first and the second operand of VectorShuffle.
+  std::vector<const analysis::Constant*> concatenated_components;
+  for (uint32_t i : {1, 2}) {
+    assert(inst->GetInOperand(i).type == SPV_OPERAND_TYPE_ID &&
+           "The vector operand must have a SPV_OPERAND_TYPE_ID type");
+    uint32_t operand_id = inst->GetSingleWordInOperand(i);
+    analysis::Constant* operand_const = FindRecordedConst(operand_id);
+    if (!operand_const) return nullptr;
+    const analysis::Type* operand_type = operand_const->type();
+    assert(operand_type->AsVector() &&
+           "The first two operand of VectorShuffle must be of vector type");
+    if (analysis::VectorConstant* vec_const =
+            operand_const->AsVectorConstant()) {
+      // case 1: current operand is a non-null vector constant.
+      concatenated_components.insert(concatenated_components.end(),
+                                     vec_const->GetComponents().begin(),
+                                     vec_const->GetComponents().end());
+    } else if (operand_const->AsNullConstant()) {
+      // case 2: current operand is a null vector constant. Create a temporary
+      // null scalar constant as the component.
+      const analysis::Type* component_type =
+          operand_type->AsVector()->element_type();
+      null_constants_to_build.emplace_back(CreateConst(component_type, {}));
+      // Append the null scalar consts to the concatenated components
+      // vector.
+      concatenated_components.insert(concatenated_components.end(),
+                                     operand_type->AsVector()->element_count(),
+                                     null_constants_to_build.back().get());
+    } else {
+      // no other valid cases
+      return nullptr;
+    }
+  }
+  // Create null component constants if there are any.
+  for (auto& c : null_constants_to_build) {
+    BuildInstructionAndAddToModule(std::move(c), pos);
```

Comment:
Hmm, yes, right, and the two vector here should have the same component type? What if one of them is vector of int another is vector of uint?